server/datastore/mysql/software.go

```diff
@@ -85,17 +75,199 @@ func (ds *Datastore) UpdateHostSoftware(ctx context.Context, hostID uint, softwa
 	})
 }
 
+func (ds *Datastore) UpdateHostSoftwareInstalledPaths(
+	ctx context.Context,
+	hostID uint,
+	reported map[string]string,
+) error {
+	stored, err := ds.getHostSoftwareWithInstalledPaths(ctx, hostID)
+	if err != nil {
+		return err
+	}
+
+	toI, toD, err := hostSoftwareInstalledPathsDelta(hostID, reported, stored)
+	if err != nil {
+		return err
+	}
+	if len(toI) == 0 && len(toD) == 0 {
+		// Nothing to do ...
+		return nil
+	}
+
+	return ds.withRetryTxx(ctx, func(tx sqlx.ExtContext) error {
+		if err := deleteHostSoftwareInstalledPaths(ctx, tx, toD); err != nil {
+			return err
+		}
+
+		if err := insertHostSoftwareInstalledPaths(ctx, tx, toI); err != nil {
+			return err
+		}
+
+		return nil
+	})
+}
+
+func (ds *Datastore) getHostSoftwareWithInstalledPaths(
+	ctx context.Context,
+	hostID uint,
+) (
+	map[string]fleet.HostSoftwareInstalledPath,
+	error,
+) {
+	// This query needs to include all columns used in 'ToUniqueStr' method
+	stmt := `
+		SELECT
+			s.id,
+			s.name,
+			s.version,
+			s.source,
+			s.bundle_identifier,
+			s.release,
+			s.vendor,
+			s.arch,
+			COALESCE(hsip.installed_path, '') AS installed_path
+		FROM software s
+			INNER JOIN host_software hs ON hs.host_id = ? AND hs.software_id = s.id
+			LEFT JOIN host_software_installed_paths hsip ON hsip.host_id = ? AND s.id = hsip.software_id
+		ORDER BY s.id
+	`
+	args := []interface{}{hostID, hostID}
+
+	var qResult []struct {
+		fleet.Software
+		InstalledPath string `db:"installed_path"`
+	}
+	if err := sqlx.SelectContext(ctx, ds.reader, &qResult, stmt, args...); err != nil {
+		return nil, err
+	}
+
+	result := make(map[string]fleet.HostSoftwareInstalledPath, len(qResult))
+	for _, s := range qResult {
+		result[s.ToUniqueStr()] = fleet.HostSoftwareInstalledPath{
+			HostID:        hostID,
+			SoftwareID:    s.ID,
+			InstalledPath: s.InstalledPath,
+		}
+	}
+	return result, nil
+}
+
+func hostSoftwareInstalledPathsDelta(
+	hostID uint,
+	reported map[string]string,
+	stored map[string]fleet.HostSoftwareInstalledPath,
+) (
+	toInsert []fleet.HostSoftwareInstalledPath,
+	toDelete []fleet.HostSoftwareInstalledPath,
+	err error,
+) {
+	for unqStr, entry := range stored {
+		_, ok := reported[unqStr]
+		// Anything not reported should be deleted
```

Comment:
Would this basically support the scenario of cleaning up `host_software_installed_paths` when a user uninstalls a software from their device?